arcontrol2nwb/converter_cli.py

```diff
@@ -127,88 +262,633 @@ def c_create(MAT):
+                 DynamicTableRegion(name="state_type",
+                                    data=state_types,
+                                    table=state_types_table,
+                                    description="The type of state that occurred on each trial. "
+                                                "This is represented as a reference to a row "
+                                                "of the StateTypesTable.")]
+    )
+    nwbfile.add_acquisition(states_table)
+    return states_table
+
+
+def __add_ndx_beadl_trials(nwbfile: NWBFile,
+                           states_table: StatesTable,
+                           events_table: EventsTable,
+                           actions_table: ActionsTable,
+                           cs_events: dict,
+                           time_offset: float):
+    """
+    Add the TrialsTable data to the NWBFile using the NDX BEADL extension
+
+    :param nwbfile: The NWBFile object ot add the data to
+    :param states_table: The StateTable from the nwbfile
+    :param events_table: The EventsTable form the nwbfile
+    :param actions_table: The ActionsTable from the nwbfile
+    :param cs_events: Dictionary with all control states
+    :param time_offset: Offset in seconds to apply to timesteps to align with the reference time of the NWB file
+
+    :return: TrialsTable generate and added to the nwbfile
+    """
+    event_timestamps = events_table['timestamp']
+    action_timestamps = actions_table['timestamp']
+    state_start_times = states_table['start_time']
+
+    # create the trials table.
+    # Here we treat components in ARControl as trials
+    trials_table = TrialsTable(description="ARControl behavioral trials ",
+                               states_table=states_table,
+                               events_table=events_table,
+                               actions_table=actions_table)
+    # Extract the trials from the states
+    trials = [(state_name,
+               timerange[0] + time_offset,  # start time
+               timerange[0] + timerange[1] + time_offset  # stop_time
+               )
+              for state_name, state_times in cs_events.items()
+              for timerange in state_times
+              # I.e., if we have a component (e.g., C2) without state name (e.g., C2S1) then add
+              # the component to the trials table instead of keeping it as a state
+              if 'S' not in state_name]
+
+    # iterate through trials sorted by starting time
+    curr_event_start_index = 0
+    curr_event_end_index = 0
+    curr_action_start_index = 0
+    curr_action_end_index = 0
+    curr_state_start_index = 0
+    curr_state_end_index = 0
+    for component_name, start_time, stop_time in sorted(trials, key=lambda t: t[1]):
+        # Find the end index for the actions, events, and states within the trial
+        # TODO Multiple  events/actions/states can have the same timestamp. Check that those are assigned correctly.
```

Comment:
chenxinfeng4 in the ARControl output there are times when events have the exact same timestamp. Here I'm trying to determine which events belong to which component/trial based on the timing. However, how do I determine whether an event belongs to a given component when the timestamp is the same as start/stop of the event, i.e., is an event that starts at the same time as a component considered to be part of that component or is it the last event of the previous component. Similarly, if an event has the same time as the end of the component, should it be counted as part of the component or as the start of the next one?


https://github.com/oruebel/ArControl-convert2-nwb/blob/063e674e1cf463d475071520c075c98a494541e8/playground/2022-1113-224711.txt#L53-L54

Reply:
The original log
```
OUT1:26261 1000
C1S1:26261
C1S2:27261
C1S3:32261
C1S5:34261
%2
OUT2:37681 2000
C2S1:37681
```

Equals  to
```
C1S1:26261
OUT1:26261 1000 #belongs to C1S1
C1S2:27261
...
C2S1:37681
OUT2:37681 2000  #belongs to C2S1
```
The OUTPUT logging is always just one step ahead of the STATE logging for the sake of better close loop stimulation.  In short, if an action(output) starts in the joint of STATE1 END/ STATE2 BEGIN, it belongs to **STATE2**; if a event(input) starts in the joint of STATE1 END/STATE2 BEGIN, it belongs to **STATE1**.